payment/src/main/java/org/killbill/billing/payment/core/janitor/IncompletePaymentAttemptTask.java

```diff
@@ -31,35 +35,41 @@
 import org.killbill.billing.payment.api.PaymentApiException;
 import org.killbill.billing.payment.api.TransactionStatus;
 import org.killbill.billing.payment.core.sm.PaymentControlStateMachineHelper;
-import org.killbill.billing.payment.core.sm.PaymentStateMachineHelper;
 import org.killbill.billing.payment.core.sm.PluginControlPaymentAutomatonRunner;
 import org.killbill.billing.payment.core.sm.control.PaymentStateControlContext;
 import org.killbill.billing.payment.dao.PaymentAttemptModelDao;
 import org.killbill.billing.payment.dao.PaymentDao;
 import org.killbill.billing.payment.dao.PaymentTransactionModelDao;
 import org.killbill.billing.payment.dao.PluginPropertySerializer;
 import org.killbill.billing.payment.dao.PluginPropertySerializer.PluginPropertySerializerException;
-import org.killbill.billing.util.callcontext.CallContext;
+import org.killbill.billing.payment.plugin.api.PaymentTransactionInfoPlugin;
+import org.killbill.billing.util.UUIDs;
+import org.killbill.billing.util.callcontext.CallOrigin;
 import org.killbill.billing.util.callcontext.InternalCallContextFactory;
+import org.killbill.billing.util.callcontext.UserType;
 import org.killbill.billing.util.config.definition.PaymentConfig;
 import org.killbill.billing.util.entity.Pagination;
 import org.killbill.clock.Clock;
-import org.killbill.commons.locker.GlobalLocker;
+import org.killbill.commons.locker.LockFailedException;
+import org.killbill.notificationq.api.NotificationEvent;
 import org.killbill.notificationq.api.NotificationQueue;
+import org.skife.config.TimeSpan;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Predicate;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Iterables;
 
-/**
- * Task to complete 'partially' incomplete payment attempts. Tis only matters for calls that went through PaymentControl apis.
- * <p/>
- * If the state of the transaction associated with the attempt completed, but the attempt state machine did not,
- * we rerun the retry state machine to complete the call and transition the attempt into a terminal state.
- */
-public class IncompletePaymentAttemptTask extends CompletionTaskBase<PaymentAttemptModelDao> {
+import static org.killbill.billing.payment.core.janitor.IncompletePaymentTransactionTask.TRANSACTION_STATUSES_TO_CONSIDER;
+
+// Janitor implementation for payments going through the control APIs.
+// Invoked on-the-fly or as part of the Janitor notification queue
```

Comment:
I am slightly confused what ` on-the-fly` means -- it thought if this was from the context of a notification thread, this would not be on the fly?

Reply:
`on-the-fly` refers to the payment code calling back the plugin during a `GET` operation (as opposed to doing it via the Janitor notification queue).

`IncompletePaymentAttemptTask` and `IncompletePaymentTransactionTask` do the actual Janitor logic. They are called from two places:
* The `Janitor` class, which orchestrates the notification queue and in-memory task.
* The `PaymentRefresher` and `PaymentControlAwareRefresher` classes, which proxy the `GET` APIs

(See also the implementation details on our wiki - if it's still confusing, I can update it)